core/src/main/java/org/infinispan/interceptors/impl/AsyncInvocationStage.java

```diff
@@ -61,20 +52,30 @@ public Object get() throws Throwable {
 
    @Override
    public boolean isDone() {
+      lockHandlers();
```

Comment:
Yeah, it's not really necessary, because we can't protect from the user calling `s1 = stage.compose(...); s2 = stage.compose(...);` and having s2's handler receive the return value from s1's handler anyway. I first intended freezing to throw exceptions when compose was called again, but then I changed it to be used mostly instead of resizing the deque.